android-kotlin/app/src/main/java/com/example/msaldemo/MainActivity.kt

```diff
@@ -0,0 +1,184 @@
+package com.example.msaldemo
+
+// Required dependencies for Android
+import androidx.appcompat.app.AppCompatActivity
+import android.os.Bundle
+import android.widget.Button
+import android.widget.TextView
+
+// Required dependencies for MSAL
+import com.microsoft.identity.client.*
+import com.microsoft.identity.client.exception.MsalException
+
+// Required dependencies for Volley - Used for making HTTP requests to Graph
+import com.android.volley.toolbox.JsonObjectRequest
+import com.android.volley.toolbox.Volley
+
+// Required dependency for working with JSON objects
+import org.json.JSONObject
+
+class MainActivity : AppCompatActivity() {
+    private var msalApplication: ISingleAccountPublicClientApplication? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.content_main)
+
+        // Creates a Public Application using MSAL configuration for single account access
+        PublicClientApplication.createSingleAccountPublicClientApplication(
+            this,
+            R.raw.msal_auth_config,
+            object : IPublicClientApplication.ISingleAccountApplicationCreatedListener {
+                override fun onCreated(application: ISingleAccountPublicClientApplication) {
+                    msalApplication = application
+                    loadAccount()
+                }
+
+                override fun onError(exception: MsalException?) {
+                    print(exception.toString())
+                }
+            })
+    }
+
+    private fun initializeUI() {
+        val btnSignIn: Button = findViewById(R.id.btn_signIn)
+        val btnRemoveAccount: Button = findViewById(R.id.btn_signOut)
+        val btnCallGraph: Button = findViewById(R.id.btn_callGraph)
+
+        // Full directory URL, in the form of https://login.microsoftonline.com/<tenant>
+        val authority = ""
+
+        btnSignIn.setOnClickListener{
+            msalApplication!!.signIn(this, "", arrayOf("user.read"), getAuthInteractiveCallback())
+        }
+
+        btnRemoveAccount.setOnClickListener{
+            // Sign the user out
+            msalApplication!!.signOut(object :
+                ISingleAccountPublicClientApplication.SignOutCallback {
+                override fun onSignOut() {
+                    updateUI(null)
+                    signOut()
+                }
+
+                override fun onError(exception: MsalException) {
+                    displayError(exception)
+                }
+            })
+        }
+
+        btnCallGraph.setOnClickListener{
+            // Attempt to use a token from the cache.
+            // If it has expired, a new token will be requested.
+            msalApplication!!.acquireTokenSilentAsync(
+                arrayOf("user.read"),
+                authority,
+                getAuthInteractiveCallback()
+            )
+        }
```

Comment:
Might be good to do the 'override fun onError(exception: MsalException) { displayError(...)} here, as ther eis a chance this will throw a MsalException if the token cannot be refreshed for some reason (like AD is forcing an interactive experience due to Conditional Access Policies or something). Thoughts?

Reply:
Yeah, my message was about the "silent" call, which doesn't make an HTTP request until the token needs to be refreshed. The error you are receiving there I think is coming from the prior msal client usage. I was suggesting that the "silent" call support basic error handling as well. But it's optional, so no big deal either way.